### PR TITLE
a quick enemy system

### PR DESCRIPTION
**Pull request by ChaoticTechSupport**
- adds enemy movement
 - adds enemy's

### CODE REVIEW

File: Scenes/Entities/Crawler/crawler.gd [hunk 1]
```diff
@@ -0,0 +1,59 @@
+extends Node2D
+class_name EnemyNode
```
Comment (StrawberrySmoothieDev) [inline Scenes/Entities/Crawler/crawler.gd:2]: Note: I'd make players, NPCs and enemies all inherit some base class, for example Actor. It's a preferential thing but lets me put some basic stuff like damage and other functions in one place.
Also, when you're making a base class (like this one) generally don't give it functionality, instead make a base EnemyNode class and and CrawlerEnemy class that extends it with the functionality. Also
dockfunni

File: Scenes/Entities/Crawler/crawler.gd [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+# --- Public Properties ---
+var current_cell: Vector2i                  ## Current cell position of the enemy in the grid.
+var target_cell: Vector2                    ## Target cell position for movement.
+var path: Array = []                        ## Array to hold the calculated path for movement.
+var path_index: int = 0                     ## Current index in the path array.
+var move: bool = false                      ## Flag indicating if the enemy is currently performing an action.
+
+# --- Private Properties ---
+@onready var _actions: int = actions
```
Comment (StrawberrySmoothieDev) [inline Scenes/Entities/Crawler/crawler.gd:16]: Would probably rename this to _current_actions. Also probably should be included in Actor base class.

File: Scenes/Entities/Crawler/crawler.gd [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		path.clear()
+		path_index = 0
+		move = false
+		if _actions == 0:
+			_actions = actions
+			turn_ended.emit()
+	
+# --- Custom Methods ---
+## Resets the enemy's actions for a new turn.
+func start_turn():
```
Comment (StrawberrySmoothieDev) [inline Scenes/Entities/Crawler/crawler.gd]: Another thing that could be moved to the actor class. We can do this in another pr tho.

File: Scenes/World.tscn [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
...z/AAAAAAAAAgD///3/AAAAAAAAAgD//wYAAAAAAAAAAgAAAPv/AAAAAAAAAgAAAPz/AAAAAAAAAQAAAP3/AAAAAAAAAQAAAAYAAAAAAAAAAgABAPv/AAAAAAAAAgABAPz/AAAAAAAAAQABAP3/AAAAAAAAAQABAAYAAAAAAAAAAgACAPv/AAAAAAAAAgACAPz/AAAAAAAAAQACAP3/AAAAAAAAAQACAAYAAAAAAAAAAgADAPv/AAAAAAAAAgADAPz/AAAAAAAAAQADAP3/AAAAAAAAAQADAAYAAAAAAAAAAgAEAPv/AAAAAAAAAgAEAPz/AAAAAAAAAQAEAP3/AAAAAAAAAQAEAAYAAAAAAAAAAgAFAPv/AAAAAAAAAgAFAPz/AAAAAAAAAQAFAP3/AAAAAAAAAQAFAAYAAAAAAAAAAgAGAPv/AAAAAAAAAgAGAPz/AAAAAAAAAQAGAP3/AAAAAAAAAQAGAAYAAAAAAAAAAgAHAPv/AAAAAAAAAgAHAPz/AAAAAAAAAQAHAP3/AAAAAAAAAQAHAAYAAAAAAAAAAgAIAPv/AAAAAAAAAgAIAPz/AAAAAAAAAQAIAP3/AAAAAAAAAQAIAAYAAAAAAAAAAgAJAPv/AAAAAAAAAgAJAPz/AAAAAAAAAQAJAP3/AAAAAAAAAQAJAAYAAAAAAAAAAgAKAPv/AAAAAAAAAgAKAPz/AAAAAAAAAQAKAP3/AAAAAAAAAQAKAAYAAAAAAAAAAgALAPv/AAAAAAAAAgALAPz/AAAAAAAAAQALAP3/AAAAAAAAAQALAAYAAAAAAAAAAgAMAPv/AAAAAAAAAgAMAPz/AAAAAAAAAQAMAP3/AAAAAAAAAQAMAAYAAAAAAAAAAgANAPv/AAAAAAAAAgANAPz/AAAAAAAAAQANAP3/AAAAAAAAAQANAAYAAAAAAAAAAgAOAPv/AAAAAAAAAgAOAPz/AAAAAAAAAQAOAP3/AAAAAAAAAQAOAAYAAAAAAAAAAgAPAPv/AAAAAAAAAgAPAPz/AAAAAAAAAQAPAP3/AAAAAAAAAQAPAAYAAAAAAAAAAgAQAPv/AAAAAAAAAgAQAPz/AAAAAAAAAQAQAP3/AAAAAAAAAQAQAAYAAAAAAAAAAgARAPv/AAAAAAAAAgARAPz/AAAAAAAAAgARAP3/AAAAAAAAAgARAAYAAAAAAAAAAgA=")
```
Comment (StrawberrySmoothieDev) [inline Scenes/World.tscn:38]: so real bestie

File: Scripts/Global/UtilityFunctions.gd [hunk 5]
```diff
@@ -15,3 +15,10 @@ static func get_all_Children(root_node, stop_class = null) -> Array[Node]: ## Ge
 
 static func in_map(cell: Vector2): ## detects if position is in map (for error handeling)
 	return global.map.local_to_map(cell) in global.map.get_used_cells()
+
+## Note: This method ignores class_name declarations. If this object's script has defined a class_name, the base, built-in class name is returned instead.
+static func get_from_type(children: Array, type_name: String) -> int:
```
Comment (StrawberrySmoothieDev) [inline Scripts/Global/UtilityFunctions.gd:20]: Don't need this. The only implementation I can see here is redundant, see my comments there.

File: Scripts/Player/Player.gd [hunk 6]
```diff
@@ -15,32 +15,37 @@ var path: Array = []                        ## Array to hold the calculated path
...
 var highlight_path: PackedVector2Array = [] ## Array to hold the positions to highlight on the path.
 var selected_object: InteractableObject     ## Reference to the interactable object currently selected by the player.
+var camera: Camera2D                        ## Refrence to Camera
 var action: bool = false:                   ## Flag indicating if the player is currently performing an action.
 	set(value):
 		action = value
 		queue_redraw()
 
+# --- Private Properties ---
+@onready var _actions: int = actions
```
Comment (StrawberrySmoothieDev) [inline Scripts/Player/Player.gd:25]: See previous comment.

File: Scripts/Player/Player.gd [hunk 7]
```diff
@@ -15,32 +15,37 @@ var path: Array = []                        ## Array to hold the calculated path
...
+
 # --- Signals ---
 signal moved(position: Vector2i)            ## Signal emitted when the player’s position is updated.
-
+signal turn_ended                           ## Signal emitted when the player's turn ends
 # --- Built-in Callbacks ---
 # Called when the node is added to the scene. Sets up camera limits and connects the resize signal.
 func _ready() -> void:
-	
+	camera = get_child(UtilityFunctions.get_from_type(get_children(), "Camera2D"))
```
Comment (StrawberrySmoothieDev) [inline Scripts/Player/Player.gd:33]: Instead may be more prudent to simply give an @onready variable for the camera, or 
`for i in get_children():
    if i is Camera2D:
          camera = i`
This is mostly unnecessary.

File: Scripts/Player/Player.gd [hunk 8]
```diff
@@ -15,32 +15,37 @@ var path: Array = []                        ## Array to hold the calculated path
...
 	global.player = self
-	$Camera2D.limit_left = global.map.get_used_rect().position.x * 128
-	$Camera2D.limit_top = global.map.get_used_rect().position.y * 128
-	$Camera2D.limit_right = (global.map.get_used_rect().position.x + global.map.get_used_rect().size.x) * 128
-	$Camera2D.limit_bottom = (global.map.get_used_rect().position.y + global.map.get_used_rect().size.y) * 128
+	camera.limit_left = global.map.get_used_rect().position.x * 128
```
Comment (StrawberrySmoothieDev) [inline Scripts/Player/Player.gd:35]: This is a little confusing, could you document it?

**Review by StrawberrySmoothieDev — CHANGES_REQUESTED** [review]
Looks pretty good to me. There's some tweaks I'd like made but this is some good stuff. Well done!! :3

File: Scripts/Player/Player.gd (hunk 8 above)
Comment (ChaoticTechSupport) [reply Scripts/Player/Player.gd:35]: sure! but thats nothing new. (that was already there from a previous PR)

File: Scenes/Entities/Entity.gd [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+# --- Signals ---
+# Define custom signals here if needed, e.g., for health changes or status updates.
+# signal health_changed(new_health)
+
+# --- Built-in Callbacks ---
+func _ready() -> void:
+	# Called when the node enters the scene tree for the first time.
+	# Assign the map to the entity's parent, assuming it's a TileMap layer.
+	map = get_parent()
```
Comment (StrawberrySmoothieDev) [inline Scenes/Entities/Entity.gd:26]: There are some situations where an entity might not be a child of a the tilemap. Also generally things shouldn't be children of the map but that's beyond the scope of this PR. Recommend making a group or a global autoload to keep track of the map.

**Review by StrawberrySmoothieDev — APPROVED** [review]
Looks amazing, well done as always! There's a couple things that probably should be in other PRs but it should be fine for now.

File: Scenes/Entities/Entity.gd (hunk 9 above)
Comment (ChaoticTechSupport) [reply Scenes/Entities/Entity.gd:26]: > There are some situations where an entity might not be a child of a the tilemap. Also generally things shouldn't be children of the map but that's beyond the scope of this PR. Recommend making a group or a global autoload to keep track of the map.

Okay, so a couple questions for my understanding. 1. What is wrong with entities being children of the tile map? Is there an offset issue or something else? 2. Can you give me a scenario where an entity would not be part of the tile map, so i know what to prepare for? (i tried to make the entire system easily configureable via tilemap, becuse everything is going to be procedurly generated.)

Comment (StrawberrySmoothieDev) [reply Scenes/Entities/Entity.gd:26]: I find that when I have a lot of nodes in a scene I often group them under other nodes to make hiding them and filtering them easier, for example I've got a node that holds all the players, all the ship systems, all the props, etc that way I can mass operate on them without much boilerplate. It also means we can move the map around if we for some reason want to do that. If we're making the player a scene tile then that should be ok, not sure if I agree w/that implementation but idk if that's what we're doing tbh.

Comment (ChaoticTechSupport) [reply Scenes/Entities/Entity.gd:26]: > I find that when I have a lot of nodes in a scene I often group them under other nodes to make hiding them and filtering them easier, for example I've got a node that holds all the players, all the ship systems, all the props, etc that way I can mass operate on them without much boilerplate. It also means we can move the map around if we for some reason want to do that. If we're making the player a scene tile then that should be ok, not sure if I agree w/that implementation but idk if that's what we're doing tbh.

yea, sorry i just have all ways wanted to implement the player as a tile, so when i got the chance i went with it.

**Review by StrawberrySmoothieDev — APPROVED** [review]

**Review by theCodeBear — APPROVED** [review]

**Review by Dragon20C — APPROVED** [review]
Looks great, and I love you are using a base entity class that all npcs and players will inherit, good job!
